File: app/src/main/java/com/example/habithero/FirebaseHelper.java
package com.example.habithero;


import android.net.Uri;
import android.util.Log;

import com.google.firebase.Timestamp;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;


import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FirebaseHelper {

        private FirebaseFirestore db = FirebaseFirestore.getInstance();
    private FirebaseStorage storage = FirebaseStorage.getInstance();




    // Method to fetch all habits for a specific user
        public void fetchUserHabits(String userId, final FirestoreCallback<List<Habit>> callback) {
            db.collection("users").document(userId).collection("habits")
                    .orderBy("timestamp", Query.Direction.DESCENDING)
                    .get()
                    .addOnCompleteListener(task -> {
                        if (task.isSuccessful()) {
                            List<Habit> habits = new ArrayList<>();
                            for (QueryDocumentSnapshot document : task.getResult()) {
                                Habit habit = document.toObject(Habit.class);
                                habit.setId(document.getId());
                                habits.add(habit);
                                Log.d("FirebaseHelper", "Fetched Habit ID: " + habit.getId() + ", Name: " + habit.getName());
                            }
                            callback.onCallback(habits);
                        } else {
                            Log.e("FirebaseHelper", "Error fetching habits: " + task.getException().getMessage(), task.getException());
                            callback.onError(task.getException());
                        }
                    });
        }

    public void loadSpecificCategoryData(String categoryName) {
        db.collection("Category") // Replace with your collection path
                .whereEqualTo("name", categoryName) // Use the field that identifies your category
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        for (DocumentSnapshot document : task.getResult()) {
                            if (document.exists()) {
                                String name = document.getString("name");
                                List<String> habitList = (List<String>) document.get("habit_list");

                                // Do something with the data, e.g., display in UI
                                Log.d("Specific Category", "Category Name: " + name + ", Habits: " + habitList);
                            } else {
                                Log.d("Error", "No such category");
                            }
                        }
                    } else {
                        Log.w("Error", "Error getting documents: ", task.getException());
                    }
                });
    }

    public void fetchCategories(FirestoreCallback<Map<String, Category>> callback) {
        db.collection("Category")
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Map<String, Category> categories = new HashMap<>();
                        for (QueryDocumentSnapshot document : task.getResult()) {
                            String categoryName = document.getString("name");
                            List<String> habitList = (List<String>) document.get("habit_list");
                            String iconPath = document.getString("icon"); // Fetch the path


                            categories.put(categoryName, new Category(categoryName, habitList, iconPath));
                        }
                        callback.onCallback(categories);
                    } else {
                        Log.e("FirebaseHelper", "Error fetching categories: " + task.getException().getMessage(), task.getException());
                        callback.onError(task.getException());
                    }
                });
    }





    public void fetchIconUrlForCategory(String category, FirestoreCallback<String> callback) {
        db.collection("Category")
                .whereEqualTo("name", category)
                .limit(1)
                .get()
                .addOnSuccessListener(queryDocumentSnapshots -> {
                    if (!queryDocumentSnapshots.isEmpty()) {
                        String iconUrl = queryDocumentSnapshots.getDocuments().get(0).getString("icon");//
                        Log.d("FirebaseHelper", "Fetched icon URL: " + iconUrl); // Added log statement
                        callback.onCallback(iconUrl);
                    } else {
                        callback.onError(new Exception("Category not found"));
                    }
                })
                .addOnFailureListener(e -> {
                    Log.e("FirebaseHelper", "2 Error fetching icon URL for category: " + e.getMessage(), e); // Added log statement
                    callback.onError(e);
                });
    }








    // Method to add a new habit for a user
        public void addHabit(String userId, Habit habit, final FirestoreCallback<Habit> callback) {
            // Set initial values for the habit
            habit.setTimestamp(new Timestamp(new Date()));
            habit.setCompleted(false); // Using 'completed' field

            db.collection("users").document(userId).collection("habits")
                    .add(habit)
                    .addOnSuccessListener(documentReference -> {
                        // Fetch the document ID and set it as the habit's ID
                        String habitId = documentReference.getId();
                        habit.setId(habitId);

                        Log.d("FirebaseHelper", "Habit successfully added to Firestore with ID: " + habitId);

                        // Pass the updated habit object to the callback
                        callback.onCallback(habit);
                    })
                    .addOnFailureListener(e -> {
                        Log.e("FirebaseHelper", "Error adding habit: " + e.getMessage(), e);
                        callback.onError(e);
                    });
        }







    // Method to update an existing habit
    public void updateCompletedHabit(String userId, Habit habit, final FirestoreCallback<Void> callback) {
        Log.d("FirebaseHelper", "Updating habit with ID: " + habit.getId());

        // Create a map for updating specific fields
        Map<String, Object> updates = new HashMap<>();
        updates.put("completed", habit.getCompleted());
        updates.put("streakCount", habit.getStreakCount());
        updates.put("completionCount", habit.getCompletionCount());

        // Update only the specific fields in the document
        db.collection("users").document(userId).collection("habits").document(habit.getId())
                .update(updates)
                .addOnSuccessListener(aVoid -> {
                    Log.d("FirebaseHelper", "Habit updated successfully: " + habit.getId());
                    callback.onCallback(null);
                })
                .addOnFailureListener(e -> {
                    Log.e("FirebaseHelper", "Error updating habit: " + e.getMessage(), e);
                    callback.onError(e);
                });
    }




    // Method to delete a habit
        public void deleteHabit(String userId, String habitId, final FirestoreCallback<Void> callback) {
            db.collection("users").document(userId).collection("habits").document(habitId)
                    .delete()
                    .addOnSuccessListener(aVoid -> callback.onCallback(null))
                    .addOnFailureListener(e -> callback.onError(e));
        }





    // Callback interface for Firestore operations
    public interface FirestoreCallback<T> {
        void onCallback(T result);
        void onError(Exception e);
    }


    public void saveProfileImage(String userId, byte[] imageData, FirestoreCallback<Void> callback) {
        FirebaseStorage storage = FirebaseStorage.getInstance();
        StorageReference storageRef = storage.getReference();
        StorageReference profileImageRef = storageRef.child("profile_images/" + userId + ".png");

        profileImageRef.putBytes(imageData)
                .addOnSuccessListener(taskSnapshot -> {
                    // Update Firestore document with the URL of the uploaded image
                    profileImageRef.getDownloadUrl().addOnSuccessListener(uri -> {
                        FirebaseFirestore db = FirebaseFirestore.getInstance();
                        Map<String, Object> updates = new HashMap<>();
                        updates.put("profileImageUrl", uri.toString());
                        db.collection("users").document(userId).update(updates);
                    });
                    callback.onCallback(null);
                })
                .addOnFailureListener(e -> callback.onError(e));
    }

    public void loadProfileImage(String userId, FirestoreCallback<Uri> callback) {
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        db.collection("users").document(userId).get()
                .addOnSuccessListener(documentSnapshot -> {
                    if (documentSnapshot.exists() && documentSnapshot.contains("profileImageUrl")) {
                        String imageUrl = documentSnapshot.getString("profileImageUrl");
                        callback.onCallback(Uri.parse(imageUrl));
                    } else {
                        callback.onError(new Exception("Profile image not found"));
                    }
                })
                .addOnFailureListener(e -> callback.onError(e));
    }

    public void updateProfileImageUrl(String userId, Uri newImageUrl, FirestoreCallback<Void> callback) {
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        Map<String, Object> updates = new HashMap<>();
        updates.put("profileImageUrl", newImageUrl.toString());

        db.collection("users").document(userId)
                .update(updates)
                .addOnSuccessListener(aVoid -> callback.onCallback(null))
                .addOnFailureListener(callback::onError);
    }

    //ToDO - Fetch Habits through firestore
    // Method to fetch predefined habits by category
    public void fetchPredefinedHabitsByCategory(String categoryName, FirestoreCallback<List<Habit>> callback) {
        db.collection("predefinedHabits").document(categoryName).collection("habits")
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        List<Habit> habits = new ArrayList<>();
                        for (QueryDocumentSnapshot document : task.getResult()) {
                            Habit habit = document.toObject(Habit.class);
                            habits.add(habit);
                        }
                        callback.onCallback(habits);
                    } else {
                        Log.e("FirebaseHelper", "Error fetching predefined habits: " + task.getException().getMessage(), task.getException());
                        callback.onError(task.getException());
                    }
                });
    }

    public void updateUsername(String userId, String newUsername, final FirestoreCallback<Void> callback) {
        db.collection("users").document(userId)
                .update("username", newUsername)
                .addOnSuccessListener(aVoid -> {
                    Log.d("FirebaseHelper", "Username updated successfully.");
                    callback.onCallback(null);
                })
                .addOnFailureListener(e -> {
                    Log.e("FirebaseHelper", "Error updating username: " + e.getMessage(), e);
                    callback.onError(e);
                });
    }

    // Method to load user data
    public void loadUserData(String userId, FirestoreCallback<User> callback) {
        db.collection("users").document(userId)
                .get()
                .addOnSuccessListener(documentSnapshot -> {
                    if (documentSnapshot.exists()) {
                        User user = documentSnapshot.toObject(User.class);
                        callback.onCallback(user);
                    } else {
                        callback.onError(new Exception("User not found"));
                    }
                })
                .addOnFailureListener(e -> {
                    Log.e("FirebaseHelper", "Error loading user data: " + e.getMessage(), e);
                    callback.onError(e);
                });
    }


}

File: app/src/main/java/com/example/habithero/FirebaseHelper.kt
package com.example.habithero

import android.net.Uri
import android.util.Log
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import kotlinx.coroutines.tasks.await
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class FirebaseHelper {

    private val db = FirebaseFirestore.getInstance()

    // --- Suspend Functions for Modern Architecture ---

    suspend fun fetchUserHabitsSuspend(userId: String): List<Habit> {
        return try {
            val task = db.collection("users").document(userId).collection("habits")
                .orderBy("timestamp", Query.Direction.DESCENDING)
                .get()
                .await()
            task.mapNotNull { document -> documentToHabit(document) }
        } catch (e: Exception) {
            Log.e("FirebaseHelper", "Error fetching habits via suspend function: ", e)
            throw e
        }
    }

    suspend fun getHabitSuspend(userId: String, habitId: String): Habit? {
        return try {
            val document = db.collection("users").document(userId).collection("habits").document(habitId).get().await()
            documentToHabit(document)
        } catch (e: Exception) {
            Log.e("FirebaseHelper", "Error fetching single habit via suspend function: ", e)
            throw e
        }
    }

    suspend fun fetchCategoriesSuspend(): Map<String, Category> {
        return try {
            val task = db.collection("Category").get().await()
            task.documents.associate {
                val categoryName = it.getString("name") ?: ""
                @Suppress("UNCHECKED_CAST")
                val habitList = it.get("habit_list") as? List<String> ?: emptyList()
                val iconPath = it.getString("icon") ?: ""
                categoryName to Category(categoryName, habitList, iconPath)
            }.filterKeys { it.isNotEmpty() }
        } catch (e: Exception) {
            Log.e("FirebaseHelper", "Error fetching categories via suspend function: ", e)
            throw e
        }
    }

    suspend fun addHabitSuspend(userId: String, habit: Habit): DocumentReference {
        return try {
            db.collection("users").document(userId).collection("habits").add(habit).await()
        } catch (e: Exception) {
            Log.e("FirebaseHelper", "Error adding habit via suspend function", e)
            throw e
        }
    }

    suspend fun updateHabitSuspend(userId: String, habit: Habit) {
        try {
            db.collection("users").document(userId).collection("habits").document(habit.id).set(habit).await()
        } catch (e: Exception) {
            Log.e("FirebaseHelper", "Error updating habit via suspend function", e)
            throw e
        }
    }

    suspend fun deleteHabit(userId: String, habitId: String) {
        try {
            db.collection("users").document(userId).collection("habits").document(habitId).delete().await()
        } catch (e: Exception) {
            Log.e("FirebaseHelper", "Error deleting habit: ", e)
            throw e // Re-throw to be handled by ViewModel
        }
    }

    suspend fun deleteHabitSuspend(userId: String, habitId: String) {
        try {
            db.collection("users").document(userId).collection("habits").document(habitId).delete().await()
        } catch (e: Exception) {
            Log.e("FirebaseHelper", "Error deleting habit via suspend function", e)
            throw e
        }
    }

    private fun documentToHabit(document: DocumentSnapshot): Habit? {
        if (!document.exists()) return null
        return try {
            val completionDates = parseCompletionDates(document.get("completionDates"))
            val habit = Habit(
                name = document.getString("name") ?: "",
                category = document.getString("category") ?: "",
                emoji = document.getString("emoji") ?: "",
                completionHour = (document.getLong("completionHour") ?: 21L).toInt(),
                completionMinute = (document.getLong("completionMinute") ?: 0L).toInt(),
                iconUrl = document.getString("iconUrl") ?: "",
                completionCount = completionDates.size, // This is the fix
                completionDates = completionDates,
                reminderEnabled = document.getBoolean("reminderEnabled") ?: false,
                reminderTimeMinutes = (document.getLong("reminderTimeMinutes") ?: 15L).toInt(),
                sourceChallengeId = document.getString("sourceChallengeId"),
                sourceTemplateId = document.getString("sourceTemplateId"),
                // Deprecated fields, kept for Firestore compatibility
                completed = document.getBoolean("completed") ?: false,
                timestamp = document.getTimestamp("timestamp") ?: Timestamp.now()
            )
            habit.id = document.id
            habit
        } catch (e: Exception) {
            Log.e("FirebaseHelper", "Error parsing habit ${document.id}", e)
            null
        }
    }

    private fun parseCompletionDates(data: Any?): List<Date> {
        if (data !is List<*>) return emptyList()
        val dates = mutableListOf<Date>()
        val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

        for (item in data) {
            when (item) {
                is Timestamp -> dates.add(item.toDate())
                is String -> {
                    try {
                        sdf.parse(item)?.let { dates.add(it) }
                    } catch (e: ParseException) {
                        Log.e("FirebaseHelper", "Failed to parse old date format: $item", e)
                    }
                }
            }
        }
        return dates
    }
}
